Libraries/acdriver/Sources/CompileActionImageSet.cpp

```diff
@@ -221,117 +210,222 @@ static bool readPNGFile(std::vector<unsigned char> &contents, std::string filena
-            renditionIdiom = car_attribute_identifier_idiom_value_car;
-            break;
+            return car_attribute_identifier_idiom_value_car;
+        default:
+            return car_attribute_identifier_idiom_value_universal;
+    }
+}
+
+static enum car_rendition_value_layout
+LayoutForResizingAndCenterMode(xcassets::Resizing::Mode resizingMode, xcassets::Resizing::Center::Mode centerMode)
+{
+    switch (resizingMode) {
+        case xcassets::Resizing::Mode::ThreePartHorizontal:
+            switch (centerMode) {
+                case xcassets::Resizing::Center::Mode::Tile:
+                    return car_rendition_value_layout_three_part_horizontal_tile;
+                case xcassets::Resizing::Center::Mode::Stretch:
+                    return car_rendition_value_layout_three_part_horizontal_scale;
+                default:
+                    return car_rendition_value_layout_three_part_horizontal_tile;
+            }
+        case xcassets::Resizing::Mode::ThreePartVertical:
+            switch (centerMode) {
+                case xcassets::Resizing::Center::Mode::Tile:
+                    return car_rendition_value_layout_three_part_vertical_tile;
+                case xcassets::Resizing::Center::Mode::Stretch:
+                    return car_rendition_value_layout_three_part_vertical_scale;
+                default:
+                    return car_rendition_value_layout_three_part_vertical_tile;
+            }
+        case xcassets::Resizing::Mode::NinePart:
+            switch (centerMode) {
+                case xcassets::Resizing::Center::Mode::Tile:
+                    return car_rendition_value_layout_nine_part_tile;
+                case xcassets::Resizing::Center::Mode::Stretch:
+                    return car_rendition_value_layout_nine_part_scale;
+                default:
+                    return car_rendition_value_layout_nine_part_tile;
+            }
+        default: abort();
+    }
+}
+
+static std::vector<car::Rendition::Slice>
+SlicesForResizingModeAndCapInsets(uint32_t width, uint32_t height, xcassets::Resizing::Mode resizingMode, ext::optional<xcassets::Insets> const &capInsets)
+{
+    double topCapInset = capInsets ? capInsets->top().value_or(0) : 0;
+    double leftCapInset = capInsets ? capInsets->left().value_or(0) : 0;
+    double bottomCapInset = capInsets ? capInsets->bottom().value_or(0) : 0;
+    double rightCapInset = capInsets ? capInsets->right().value_or(0) : 0;
+
+    uint32_t leftWidth = static_cast<uint32_t>(leftCapInset);
+    uint32_t centerWidth = static_cast<uint32_t>(width - (leftCapInset + rightCapInset));
+    uint32_t rightWidth = static_cast<uint32_t>(rightCapInset);
+
+    uint32_t topHeight = static_cast<uint32_t>(topCapInset);
+    uint32_t centerHeight = static_cast<uint32_t>(height - (topCapInset + bottomCapInset));
+    uint32_t bottomHeight = static_cast<uint32_t>(bottomCapInset);
+
+    uint32_t topVerticalOffset = static_cast<uint32_t>(height - bottomCapInset);
```

Comment:
This is wrong, it should be:
uint32_t topVerticalOffset = static_cast<uint32_t>(height - topCapInset);